services/outputhost/replicaconnection.go

```diff
@@ -186,124 +192,183 @@ func (conn *replicaConnection) readMessagesPump() {
+			conn.logger.WithField(common.TagErr, err).Error(`Error reading msg from store`)
+			go conn.close(err)
+			<-conn.closeChannel
+			return
+		}
+
+		hb.Beat()
+
+		switch rmc.GetType() {
+		case store.ReadMessageContentType_MESSAGE:
+
+			msg := rmc.GetMessage()
+			msgSeqNum := common.SequenceNumber(msg.Message.GetSequenceNumber())
+			msgAddr := msg.GetAddress()
+
+			// XXX: Sequence number check to make sure we get monotonically increasing
+			// sequence number.
+			// We just log and move forward
+			// XXX: Note we skip the first message check here because we can start from
+			// a bigger sequence number in case of restarts
+			if conn.extCache.destType == shared.DestinationType_TIMER {
+
+				// T471157 For timers, do not signal discontinuities to ack manager, since discontinuities are frequent
+				msgSeqNum = 0
+
+			} else if lastSeqNum+1 != int64(msgSeqNum) {
+
+				// FIXME: add metric to help alert this case
+				expectedSeqNum := 1 + lastSeqNum
+				skippedMessages := int64(msgSeqNum) - lastSeqNum
+
+				conn.logger.WithFields(bark.Fields{
+					"msgSeqNum":       msgSeqNum,
+					"expectedSeqNum":  expectedSeqNum,
+					"skippedMessages": skippedMessages,
+				}).Error("sequence number out of order")
 			}
 
-			hb.Beat()
+			// update the lastSeqNum to this value
+			lastSeqNum = msg.Message.GetSequenceNumber()
 
-			switch rmc.GetType() {
-			case store.ReadMessageContentType_MESSAGE:
+			// convert this to an outMessage
+			cMsg := cherami.NewConsumerMessage()
+			cMsg.EnqueueTimeUtc = msg.Message.EnqueueTimeUtc
+			cMsg.Payload = msg.Message.Payload
 
-				msg := rmc.GetMessage()
-				msgSeqNum := common.SequenceNumber(msg.Message.GetSequenceNumber())
+			cMsg.AckId = common.StringPtr(conn.extCache.ackMgr.getNextAckID(storeHostAddress(msgAddr), msgSeqNum))
 
-				// XXX: Sequence number check to make sure we get monotonically increasing
-				// sequence number.
-				// We just log and move forward
-				// XXX: Note we skip the first message check here because we can start from
-				// a bigger sequence number in case of restarts
-				if conn.extCache.destType != shared.DestinationType_TIMER {
-					if lastSeqNum+1 != int64(msgSeqNum) {
-						// FIXME: add metric to help alert this case
-						expectedSeqNum := 1 + lastSeqNum
-						skippedMessages := int64(msgSeqNum) - lastSeqNum
+			if conn.extCache.destType != shared.DestinationType_TIMER {
 
-						conn.logger.WithFields(bark.Fields{
-							"msgSeqNum":       msgSeqNum,
-							"expectedSeqNum":  expectedSeqNum,
-							"skippedMessages": skippedMessages,
-						}).Error("sequence number out of order")
-					}
-				} else {
-					// T471157 For timers, do not signal discontinuities to ack manager, since discontinuities are frequent
-					msgSeqNum = 0
+				// compute the 'visibility-time' of the message, taking into account the specified 'delay', if any.
+				visibilityTime := msg.Message.GetEnqueueTimeUtc()
```

Comment:
if the visibility time is zero, then we can neither process delay nor skipolder, so we might just skip this whole if-block starting at line 269. That would also be a good place to call out that it may be zero, and specify the policy of pass-thru in that case.